proxy_connect.go

```diff
@@ -0,0 +1,178 @@
+package proxy
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/getlantern/errors"
+	"github.com/getlantern/hidden"
+	"github.com/getlantern/idletiming"
+	"github.com/getlantern/netx"
+	"github.com/getlantern/ops"
+)
+
+// BufferSource is a source for buffers used in reading/writing.
+type BufferSource interface {
+	Get() []byte
+	Put(buf []byte)
+}
+
+// CONNECT returns a Handler that processes CONNECT requests.
+//
+// defaultPort: specifies which port to use if none was found in the initial
+// request
+//
+// idleTimeout: if specified, lets us know to include an appropriate
+// KeepAlive: timeout header in the CONNECT response
+//
+// bufferSource: specifies a BufferSource, leave nil to use default
+//
+// dial: the function that's used to dial upstream
+func CONNECT(
+	defaultPort int,
```

Comment:
It's also what the standard library do https://golang.org/src/net/http/transport.go#L1933